llama.cpp

```diff
@@ -12459,6 +12459,122 @@ int32_t llama_token_to_piece(const struct llama_model * model, llama_token token
     return 0;
 }
 
+int32_t llama_chat_apply_template_internal(std::string &dest, std::string chat_template, std::vector<const llama_chat_message *> conversation, bool add_ass);
+
+// trim whitespace from the beginning and end of a string
+static std::string trim(const std::string & str) {
+    size_t start = 0;
+    size_t end = str.size();
+    while (start < end && isspace(str[start])) {
+        start += 1;
+    }
+    while (end > start && isspace(str[end - 1])) {
+        end -= 1;
+    }
+    return str.substr(start, end - start);
+}
+
+// Simple version of "llama_apply_chat_template" that only works with strings
+// This function uses heuristic checks to determine commonly used template. It is not a jinja parser.
+int32_t llama_chat_apply_template_internal(std::string &dest, std::string chat_template, std::vector<const llama_chat_message *> conversation, bool add_ass) {
```

Comment:
Put the const args in front and pass by ref:

```suggestion
int32_t llama_chat_apply_template_internal(
    const std::string & chat_template,
    const std::vector<const llama_chat_message *> & chat, 
    std::string & dest, bool add_ass) {
```

The terms `chat` and `conversation` seem conflated. Propose to use `chat` universally (apply to other places too)

Reply:
I changed all the occurrences of `conversation` and `msg` to `chat` in this commit: https://github.com/ggerganov/llama.cpp/pull/5538/commits/73fbd6790110abb8790fb5d45febb0c32a1a29c9